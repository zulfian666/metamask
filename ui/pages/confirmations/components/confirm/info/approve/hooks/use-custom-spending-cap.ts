import { useState } from 'react';

export const useCustomSpendingCap = () => {
  const [customSpendingCap, setCustomSpendingCap] = useState('');

  return { customSpendingCap, setCustomSpendingCap };
};
