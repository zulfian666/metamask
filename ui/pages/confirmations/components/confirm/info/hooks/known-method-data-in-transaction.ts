import { TransactionMeta } from '@metamask/transaction-controller';
import { useDispatch, useSelector } from 'react-redux';
import {
  getKnownMethodData,
  use4ByteResolutionSelector,
} from '../../../../../../selectors';
import { getContractMethodData } from '../../../../../../store/actions';

export const useKnownMethodDataInTransaction = (
  currentConfirmation: TransactionMeta,
) => {
  const dispatch = useDispatch();
  const use4ByteResolution = useSelector(use4ByteResolutionSelector);
  const transactionData = currentConfirmation?.txParams?.data;
  if (use4ByteResolution && transactionData) {
    dispatch(getContractMethodData(currentConfirmation.txParams.data));
  }
  const knownMethodData =
    useSelector((state) => getKnownMethodData(state, transactionData)) || {};
  return { knownMethodData };
};
