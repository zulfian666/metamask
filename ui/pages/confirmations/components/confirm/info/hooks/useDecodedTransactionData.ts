import { Hex } from '@metamask/utils';
import { useSelector } from 'react-redux';
import { TransactionMeta } from '@metamask/transaction-controller';
import {
  AsyncResult,
  useAsyncResult,
} from '../../../../../../hooks/useAsyncResult';
import { decodeTransactionData } from '../../../../../../store/actions';
import { DecodedTransactionDataResponse } from '../../../../../../../shared/types/transaction-decode';
import { currentConfirmationSelector } from '../../../../selectors';

export function useDecodedTransactionData(): AsyncResult<
  DecodedTransactionDataResponse | undefined
> {
  const currentConfirmation = useSelector(currentConfirmationSelector) as
    | TransactionMeta
    | undefined;

  const chainId = currentConfirmation?.chainId as Hex;
  const contractAddress = currentConfirmation?.txParams?.to as Hex;
  const transactionData = currentConfirmation?.txParams?.data as Hex;

  return useAsyncResult(async () => {
    if (!transactionData?.length) {
      return undefined;
    }

    return await decodeTransactionData({
      transactionData,
      chainId,
      contractAddress,
    });
  }, [transactionData, chainId, contractAddress]);
}
