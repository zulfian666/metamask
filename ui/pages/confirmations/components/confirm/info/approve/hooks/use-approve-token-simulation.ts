import { TransactionMeta } from '@metamask/transaction-controller';
import { BigNumber } from 'bignumber.js';
import { useSelector } from 'react-redux';
import { getIntlLocale } from '../../../../../../../ducks/locale/locale';
import { useDecodedTransactionData } from '../../hooks/useDecodedTransactionData';
import { useIsNFT } from './use-is-nft';

export const UNLIMITED_MSG = 'UNLIMITED MESSAGE';

const UNLIMITED_THRESHOLD = 10 ** 15;

function isSpendingCapUnlimited(decodedSpendingCap: number) {
  return decodedSpendingCap >= UNLIMITED_THRESHOLD;
}

export const useApproveTokenSimulation = (
  transactionMeta: TransactionMeta,
  decimals: string,
) => {
  const locale = useSelector(getIntlLocale);

  const { isNFT, pending: isNFTPending } = useIsNFT(transactionMeta);

  const decodedResponse = useDecodedTransactionData();

  const { value, pending } = decodedResponse;

  const decodedSpendingCap = value
    ? new BigNumber(value.data[0].params[1].value)
        .dividedBy(new BigNumber(10).pow(Number(decimals)))
        .toNumber()
    : 0;

  const tokenPrefix = isNFT ? '#' : '';
  const formattedTokenNum = isNFT
    ? decodedSpendingCap
    : new Intl.NumberFormat(locale).format(decodedSpendingCap);

  let tokenAmount;
  if (!isNFT && isSpendingCapUnlimited(decodedSpendingCap)) {
    tokenAmount = UNLIMITED_MSG;
  } else {
    tokenAmount = `${tokenPrefix}${formattedTokenNum}`;
  }

  return {
    tokenAmount,
    formattedTokenNum,
    value,
    pending: pending || isNFTPending,
  };
};
