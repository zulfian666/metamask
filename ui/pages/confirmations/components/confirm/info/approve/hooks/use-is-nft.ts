import { TransactionMeta } from '@metamask/transaction-controller';
import { TokenStandard } from '../../../../../../../../shared/constants/transaction';
import { useAsyncResult } from '../../../../../../../hooks/useAsyncResult';
import { getTokenStandardAndDetails } from '../../../../../../../store/actions';

export const useIsNFT = (
  transactionMeta: TransactionMeta,
): { isNFT: boolean; pending: boolean } => {
  const { value, pending } = useAsyncResult(async () => {
    return await getTokenStandardAndDetails(
      transactionMeta?.txParams?.to as string,
    );
  }, [transactionMeta]);

  const isNFT = value?.standard !== TokenStandard.ERC20;

  return { pending, isNFT };
};
