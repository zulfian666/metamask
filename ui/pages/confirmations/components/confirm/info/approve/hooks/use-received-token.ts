import { TransactionMeta } from '@metamask/transaction-controller';
import { useSelector } from 'react-redux';

import { normalizeSafeAddress } from '../../../../../../../../app/scripts/lib/multichain/address';
import { useAccountTotalFiatBalance } from '../../../../../../../hooks/useAccountTotalFiatBalance';
import { getSelectedAccount } from '../../../../../../../selectors';
import { useConfirmContext } from '../../../../../context/confirm';

export type TokenWithBalance = {
  address: string;
  balance: string;
  balanceError: unknown;
  decimals: number;
  image: unknown;
  isERC721: unknown;
  string: string;
  symbol: string;
};

export const useReceivedToken = () => {
  const { currentConfirmation: transactionMeta } = useConfirmContext() as {
    currentConfirmation: TransactionMeta;
  };

  const selectedAccount = useSelector(getSelectedAccount);

  const { tokensWithBalances } = useAccountTotalFiatBalance(
    selectedAccount,
    true,
  ) as { tokensWithBalances: TokenWithBalance[] };

  const receivedToken = tokensWithBalances.find(
    (token) =>
      normalizeSafeAddress(transactionMeta.txParams.to as string) ===
      normalizeSafeAddress(token.address),
  );

  return { receivedToken };
};
