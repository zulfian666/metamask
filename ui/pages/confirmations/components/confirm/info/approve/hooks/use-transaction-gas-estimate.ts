import { TransactionMeta } from '@metamask/transaction-controller';
import { useAsyncResult } from '../../../../../../../hooks/useAsyncResult';
import { estimateGas } from '../../../../../../../store/actions';

export const useTransactionGasEstimate = (
  transactionMeta: TransactionMeta,
  customTxParamsData: string,
  customSpendingCap: string,
) => {
  const { value: estimatedGasLimit } = useAsyncResult(async () => {
    return await estimateGas({
      from: transactionMeta.txParams.from,
      to: transactionMeta.txParams.to,
      value: transactionMeta.txParams.value,
      data: customTxParamsData,
    });
  }, [customTxParamsData, customSpendingCap]);

  return { estimatedGasLimit };
};
