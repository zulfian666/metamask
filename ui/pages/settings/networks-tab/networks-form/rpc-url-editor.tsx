import React, { useRef, useState } from 'react';
import classnames from 'classnames';
import { useDispatch } from 'react-redux';
import {
  Box,
  ButtonIcon,
  ButtonIconSize,
  Icon,
  IconName,
  IconSize,
  Popover,
  PopoverPosition,
  Text,
} from '../../../../components/component-library';
import {
  AlignItems,
  BackgroundColor,
  BorderColor,
  BorderRadius,
  Display,
  IconColor,
  JustifyContent,
  TextColor,
  TextVariant,
} from '../../../../helpers/constants/design-system';
import { useI18nContext } from '../../../../hooks/useI18nContext';

import { showModal, toggleNetworkMenu } from '../../../../store/actions';

export const RpcUrlEditor = ({
  currentRpcUrl,
  onRpcUrlAdd,
  onRpcSelected,
  dummyRpcUrls = [],
}: {
  currentRpcUrl: string;
  onRpcUrlAdd: () => void;
  onRpcSelected: (url: string) => void;
  dummyRpcUrls: { url: string; selected: boolean }[];
}) => {
  const t = useI18nContext();
  const dispatch = useDispatch();
  const rpcDropdown = useRef(null);
  const [isDropdownOpen, setIsDropdownOpen] = useState(false);
  const [currentRpcEndpoint, setCurrentRpcEndpoint] = useState(currentRpcUrl);

  const handleSelectRpc = (rpcEndpoint: string) => {
    onRpcSelected(rpcEndpoint);
    setCurrentRpcEndpoint(rpcEndpoint);
  };

  return (
    <>
      <Text
        className="networks-tab__rpc-header"
        marginTop={1}
        marginBottom={1}
        variant={TextVariant.bodySmBold}
      >
        {t('defaultRpcUrl')}
      </Text>
      <Box
        onClick={() => setIsDropdownOpen(!isDropdownOpen)}
        className="networks-tab__rpc-dropdown"
        display={Display.Flex}
        justifyContent={JustifyContent.spaceBetween}
        borderRadius={BorderRadius.MD}
        borderColor={BorderColor.borderDefault}
        borderWidth={1}
        padding={2}
        ref={rpcDropdown}
      >
        <Text variant={TextVariant.bodySm}>{currentRpcEndpoint}</Text>
        <ButtonIcon
          iconName={isDropdownOpen ? IconName.ArrowUp : IconName.ArrowDown}
          ariaLabel={t('defaultRpcUrl')}
          size={ButtonIconSize.Sm}
        />
      </Box>
      <Popover
        paddingTop={2}
        paddingBottom={2}
        paddingLeft={0}
        matchWidth={true}
        paddingRight={0}
        className="networks-tab__rpc-popover"
        referenceElement={rpcDropdown.current}
        position={PopoverPosition.Bottom}
        isOpen={isDropdownOpen}
        onClickOutside={() => setIsDropdownOpen(!isDropdownOpen)}
      >
        {dummyRpcUrls.map(({ url }) => (
          <Box
            alignItems={AlignItems.center}
            padding={4}
            display={Display.Flex}
            justifyContent={JustifyContent.spaceBetween}
            key={url}
            onClick={() => {
              handleSelectRpc(url);
              setIsDropdownOpen(false);
            }}
            className={classnames('networks-tab__rpc-item', {
              'networks-tab__rpc-item--selected': url === currentRpcEndpoint,
            })}
          >
            {url === currentRpcEndpoint && (
              <Box
                className="networks-tab__rpc-selected-pill"
                borderRadius={BorderRadius.pill}
                backgroundColor={BackgroundColor.primaryDefault}
              />
            )}
            <Text
              as="button"
              color={TextColor.textDefault}
              variant={TextVariant.bodySmMedium}
              backgroundColor={BackgroundColor.transparent}
              ellipsis
            >
              {url}
            </Text>
            <ButtonIcon
              marginLeft={1}
              ariaLabel={t('delete')}
              size={ButtonIconSize.Sm}
              iconName={IconName.Trash}
              color={IconColor.errorDefault}
              onClick={(e: React.MouseEvent) => {
                e.stopPropagation();
                dispatch(toggleNetworkMenu());
                dispatch(
                  showModal({
                    name: 'CONFIRM_DELETE_RPC_URL',
                  }),
                );
              }}
            />
          </Box>
        ))}
        <Box
          onClick={onRpcUrlAdd}
          padding={4}
          display={Display.Flex}
          alignItems={AlignItems.center}
          className="networks-tab__rpc-item"
        >
          <Icon
            color={IconColor.primaryDefault}
            name={IconName.Add}
            size={IconSize.Sm}
            marginRight={2}
          />
          <Text
            as="button"
            backgroundColor={BackgroundColor.transparent}
            color={TextColor.primaryDefault}
            variant={TextVariant.bodySmMedium}
          >
            {t('addRpcUrl')}
          </Text>
        </Box>
      </Popover>
    </>
  );
};

export default RpcUrlEditor;
