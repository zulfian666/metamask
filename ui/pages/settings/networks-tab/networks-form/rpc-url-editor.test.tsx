import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { useDispatch } from 'react-redux';
import { showModal, toggleNetworkMenu } from '../../../../store/actions';
import { useI18nContext } from '../../../../hooks/useI18nContext';
import { RpcUrlEditor } from './rpc-url-editor';

// Mock useDispatch
jest.mock('react-redux', () => ({
  useDispatch: jest.fn(),
}));

jest.mock('../../../../hooks/useI18nContext', () => ({
  useI18nContext: jest.fn(),
}));

describe('RpcUrlEditor', () => {
  const useDispatchMock = useDispatch as jest.Mock;
  const mockOnRpcUrlAdd = jest.fn();
  const mockOnRpcSelected = jest.fn();
  const useI18nContextMock = useI18nContext as jest.Mock;
  const mockDispatch = jest.fn();

  const defaultProps = {
    currentRpcUrl: 'https://current-rpc-url.com',
    onRpcUrlAdd: mockOnRpcUrlAdd,
    onRpcSelected: mockOnRpcSelected,
    dummyRpcUrls: [
      { url: 'https://rpc-url-1.com', selected: false },
      { url: 'https://rpc-url-2.com', selected: true },
    ],
  };

  beforeEach(() => {
    useDispatchMock.mockReturnValue(mockDispatch);
    useI18nContextMock.mockReturnValue((key: string) => key);
    jest.clearAllMocks();
  });

  it('should render correctly', () => {
    const { container } = render(<RpcUrlEditor {...defaultProps} />);
    expect(container).toMatchSnapshot();
  });

  it('should toggle the dropdown when clicked', () => {
    render(<RpcUrlEditor {...defaultProps} />);

    const dropdown = screen.getByText('https://current-rpc-url.com');
    fireEvent.click(dropdown);
    expect(screen.getByText('https://rpc-url-1.com')).toBeVisible();

    fireEvent.click(dropdown);
    expect(screen.queryByText('https://rpc-url-1.com')).not.toBeInTheDocument();
  });

  it('should call onRpcUrlAdd when "Add RPC URL" button is clicked', () => {
    render(<RpcUrlEditor {...defaultProps} />);

    const dropdown = screen.getByText('https://current-rpc-url.com');
    fireEvent.click(dropdown);

    const addButton = screen.getByText('addRpcUrl');
    fireEvent.click(addButton);

    expect(mockOnRpcUrlAdd).toHaveBeenCalled();
  });

  it('should dispatch actions when delete button is clicked', () => {
    render(<RpcUrlEditor {...defaultProps} />);

    const dropdown = screen.getByText('https://current-rpc-url.com');
    fireEvent.click(dropdown);

    const deleteButton = screen.getAllByLabelText('delete')[0];
    fireEvent.click(deleteButton);

    expect(mockDispatch).toHaveBeenCalledWith(toggleNetworkMenu());
    expect(mockDispatch).toHaveBeenCalledWith(
      showModal({ name: 'CONFIRM_DELETE_RPC_URL' }),
    );
  });
});
