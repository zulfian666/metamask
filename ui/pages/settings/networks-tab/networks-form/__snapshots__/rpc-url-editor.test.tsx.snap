// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RpcUrlEditor should render correctly 1`] = `
<div>
  <p
    class="mm-box mm-text networks-tab__rpc-header mm-text--body-sm-bold mm-box--margin-top-1 mm-box--margin-bottom-1 mm-box--color-text-default"
  >
    defaultRpcUrl
  </p>
  <div
    class="mm-box networks-tab__rpc-dropdown mm-box--padding-2 mm-box--display-flex mm-box--justify-content-space-between mm-box--rounded-md mm-box--border-color-border-default mm-box--border-width-1 box--border-style-solid"
  >
    <p
      class="mm-box mm-text mm-text--body-sm mm-box--color-text-default"
    >
      https://current-rpc-url.com
    </p>
    <button
      aria-label="defaultRpcUrl"
      class="mm-box mm-button-icon mm-button-icon--size-sm mm-box--display-inline-flex mm-box--justify-content-center mm-box--align-items-center mm-box--color-icon-default mm-box--background-color-transparent mm-box--rounded-lg"
    >
      <span
        class="mm-box mm-icon mm-icon--size-sm mm-box--display-inline-block mm-box--color-inherit"
        style="mask-image: url('./images/icons/arrow-down.svg');"
      />
    </button>
  </div>
</div>
`;
